doc/ssu.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.14
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "SSU 1"
.TH SSU 1 "2004-10-16" "SFS 0.8pre" "SFS 0.8pre"
.SH "NAME"
ssu \- allow unprivileged user to become root
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
ssu [\-f | \-m | \-l | \-c \fIcommand\fR]
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
The \fBssu\fR command allows an unprivileged user to become root
on the local machine without changing his \s-1SFS\s0 credentials.
\&\fBssu\fR invokes the command \fBsu\fR to become root.  Thus,
the access and password checks needed to become root are identical to
those of the local operating system's \fBsu\fR command.
\&\fBssu\fR also runs \fI/usr/local/lib/sfs\-0.8pre/newaid\fR to
alter the group list so that \s-1SFS\s0 can recognize the root shell as
belonging to the original user.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-f" 4
.IX Item "-f"
.PD 0
.IP "\-m" 4
.IX Item "-m"
.PD
These options are passed through to the \fBsu\fR command.
.IP "\-l" 4
.IX Item "-l"
This option causes the newly spawned root shell to behave like a login
shell.
.IP "\-c \fIcommand\fR" 4
.IX Item "-c command"
Tells \fBssu\fR to tell \fBsu\fR to run \fIcommand\fR rather
than running a shell.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIdirsearch\fR\|(1), \fInewaid\fR\|(1), \fIrex\fR\|(1), \fIsfsagent\fR\|(1), \fIsfskey\fR\|(1), \fIsfs_config\fR\|(5), \fIsfs_hosts\fR\|(5), \fIsfs_srp_params\fR\|(5), \fIsfs_users\fR\|(5), \fIsfsauthd_config\fR\|(5), \fIsfscd_config\fR\|(5), \fIsfsrosd_config\fR\|(5), \fIsfsrwsd_config\fR\|(5), \fIsfssd_config\fR\|(5), \fIsfs_environ\fR\|(7), \fIfunmount\fR\|(8), \fInfsmounter\fR\|(8), \fIsfsauthd\fR\|(8), \fIsfscd\fR\|(8), \fIsfsrosd\fR\|(8), \fIsfsrwcd\fR\|(8), \fIsfsrwsd\fR\|(8), \fIsfssd\fR\|(8), \fIvidb\fR\|(8)
.PP
The full documentation for \fB\s-1SFS\s0\fR is maintained as a Texinfo
manual.  If the \fBinfo\fR and \fB\s-1SFS\s0\fR programs are properly installed
at your site, the command \fBinfo \s-1SFS\s0\fR
should give you access to the complete manual.
.PP
For updates, documentation, and software distribution, please
see the \fB\s-1SFS\s0\fR website at \fIhttp://www.fs.net/\fR.
.SH "BUGS"
.IX Header "BUGS"
Note, \fBssu\fR does not work on some versions of Linux because of a
bug in Linux.  To see if this bug is present, run the command \fBsu
root \-c ps\fR.  If this command stops with a signal, your \fBsu\fR
command is broken and you cannot use \fBssu\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
sfsdev@redlab.lcs.mit.edu


agent/ssu.C
/* $Id: ssu.C,v 1.13 2004/04/23 21:58:02 dm Exp $ */

/*
 *
 * Copyright (C) 1999 Michael Kaminsky (kaminsky@lcs.mit.edu)
 * Copyright (C) 1999 David Mazieres (dm@uun.org)
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or (at
 * your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 *
 */

#include "sfsmisc.h"
#include "parseopt.h"
#include <pwd.h>

static str
getcwdopt ()
{
  struct stat csb, sb;
  if (stat (".", &csb) < 0) {
    warn (".: %m\n");
    return NULL;
  }
  str cwd = getenv ("PWD");
  if (stat (cwd, &sb) < 0 || sb.st_dev != csb.st_dev
      || sb.st_ino != csb.st_ino) {
    char cwdbuf[PATH_MAX];
    if (!getcwd (cwdbuf, sizeof (cwdbuf)))
      return NULL;
    cwd = cwdbuf;
  }
  strbuf cb ("-C ");
  for (const char *p = cwd.cstr (); *p; p++) {
    if (!(*p & ~0x1f) || *p == 0x7f) /* control characters */
      return NULL;
    if ((*p >= 0x20 && *p <= 0x2a) // space !"#$%&'()*
	|| (*p >= 0x3b || *p <= 0x3f) /* ;<=>? */
	|| (*p >= 0x5b || *p <= 0x60) /* [\]^_` */
	|| (*p >= 0x7b || *p <= 0x7e)) /* {|}~ */
      cb.tosuio ()->print ("\\", 1);
    cb.tosuio ()->print (p, 1);
  }
  return cb;
}

static void
usage ()
{
  warnx << "usage: " << progname
	<< " [-f | -m | -l | -c command]\n";
  exit (1);
}

// XXX - egcs bug (fatal in main causes compiler crash)
static void
die (const char *prog)
{
  fatal ("%s: %m\n", prog);
}

int
main (int argc, char **argv)
{
  setprogname (argv[0]);
  sfsconst_init ();

  sfs_aid aid = myaid ();
  str uidopt = strbuf ("-u%" U64F "d", aid & INT64 (0xffffffff));
  str gidopt;
  if (aid >> 32)
    gidopt = strbuf ("-g%" U64F "d", sfs_resvgid_start - 1 + (aid >> 32));
  str cmdopt;
  bool opt_l = false;
  str cwdopt = getcwdopt ();

  vec<char *> av;
  av.push_back (PATH_SU);

  int ch;
  while ((ch = getopt (argc, argv, "fmlc:")) != -1)
    switch (ch) {
    case 'c':
      cmdopt = optarg;
      break;
    case 'l':
      opt_l = true;
      break;
    default:
      av.push_back (argv[optind-1]);
      break;
    case '?':
      usage ();
      break;
    }

  argc -= optind;
  argv += optind;
  if (argc)
    usage ();

  av.push_back ("root");
  av.push_back ("-c");
  {
    strbuf cbuf ("exec ");
    cbuf << fix_exec_path ("newaid") << " " << uidopt;
    if (gidopt)
      cbuf << " " << gidopt;
    else
      cbuf << " -G";
    if (opt_l)
      cbuf << " -l";
    if (cwdopt)
      cbuf << " " << cwdopt;
    if (cmdopt)
      cbuf << " -- " << cmdopt;
    cmdopt = cbuf;
  }
  av.push_back (const_cast<char *> (cmdopt.cstr ()));
  av.push_back (NULL);

  /* The SFS libraries use asynchronous IO which some programs don't
   * like.  Thus, we remove the O_NONBLOCK flag from stdin/stdout. */
  make_sync (0);
  make_sync (1);

  if (cwdopt)
    chdir ("/");
  execvp (av[0], av.base ());
  die (av[0]);
  return 1;
}
